File: platforms/android/src/main/kotlin/co/trace/TraceView.kt
package co.trace

import android.content.Context
import android.graphics.Canvas
import android.graphics.RectF
import android.net.Uri
import android.util.AttributeSet
import android.view.Choreographer
import android.view.MotionEvent
import android.view.View
import io.github.shopify.skia.SkCanvas

/** Android counterpart of TraceView (SPEC §6.2). */
class TraceView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
) : View(context, attrs, defStyle) {

    sealed class Source {
        data class File(val uri: Uri) : Source()
        data class Bytes(val data: ByteArray) : Source()
        data class Streaming(val session: TraceSession) : Source()
    }

    var source: Source? = null
        set(value) { field = value; reload() }

    var listener: Listener? = null

    var highlightedIds: Set<String> = emptySet()
        set(value) { field = value; invalidate() }

    var highlightColor: Int = 0xFFFF5722.toInt()
    var highlightStrokeBoost: Float = 1.5f

    private val renderer = TraceRenderer(this)
    private var animator: Animator? = null
    private var traceFile: TraceFile? = null
    private var traceSession: TraceSession? = null

    init {
        // GPU-accelerated layer — see SPEC §6.5.
        setLayerType(LAYER_TYPE_HARDWARE, null)
    }

    internal val viewBoxBounds: RectF
        get() = traceFile?.viewBox ?: RectF(0f, 0f, width.toFloat(), height.toFloat())

    private fun reload() {
        val src = source
        if (src == null) {
            traceFile = null; traceSession = null
            invalidate(); return
        }
        try {
            when (src) {
                is Source.File -> {
                    val bytes = context.contentResolver.openInputStream(src.uri)
                        ?.use { it.readBytes() } ?: return
                    val f = TraceFile(bytes)
                    renderer.load(f)
                    traceFile = f; traceSession = null
                }
                is Source.Bytes -> {
                    val f = TraceFile(src.data)
                    renderer.load(f)
                    traceFile = f; traceSession = null
                }
                is Source.Streaming -> {
                    traceSession = src.session
                    val f = TraceFile(src.session.snapshot())
                    renderer.load(f)
                    traceFile = f
                }
            }
            startOrResetAnimator()
            invalidate()
        } catch (e: Exception) {
            android.util.Log.e("Trace", "load failed", e)
        }
    }

    override fun onDraw(canvas: Canvas) {
        // The host app's react-native-skia provides a SkCanvas bridge over
        // android.graphics.Canvas. For standalone Android apps we obtain a
        // SkCanvas from a vendored Skia surface attached to this View. In both
        // cases the drawing path is the same: per-element drawPath using the
        // identity-to-handle map.
        val skCanvas = SkCanvas.fromAndroidCanvas(canvas)
        val file = traceFile ?: return
        val vb = file.viewBox
        val sx = width / vb.width()
        val sy = height / vb.height()
        val s = minOf(sx, sy)
        val dx = (width - vb.width() * s) * 0.5f - vb.left * s
        val dy = (height - vb.height() * s) * 0.5f - vb.top * s
        skCanvas.save()
        skCanvas.translate(dx, dy)
        skCanvas.scale(s, s)
        renderer.draw(skCanvas, highlightedIds, highlightColor, highlightStrokeBoost)
        skCanvas.restore()
    }

    internal fun requestRedraw() {
        post { invalidate() }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action != MotionEvent.ACTION_UP) return super.onTouchEvent(event)
        val file = traceFile ?: return super.onTouchEvent(event)
        val vb = file.viewBox
        val sx = width / vb.width()
        val sy = height / vb.height()
        val s = minOf(sx, sy)
        val dx = (width - vb.width() * s) * 0.5f - vb.left * s
        val dy = (height - vb.height() * s) * 0.5f - vb.top * s
        val vbX = (event.x - dx) / s
        val vbY = (event.y - dy) / s
        // Hit-test routes to Rust. Skia is never consulted for identity.
        val id = file.hitTest(vbX, vbY, HitMode.VISIBLE_PAINTED)
        if (id != null) {
            listener?.onElementTap(id, event.x, event.y)
            return true
        }
        return super.onTouchEvent(event)
    }

    private fun startOrResetAnimator() {
        animator?.stop()
        animator = Animator(renderer, traceFile!!, traceSession) { id ->
            listener?.onAnimationFinished(id)
        }
        animator?.start()
    }

    interface Listener {
        fun onElementTap(id: String, x: Float, y: Float)
        fun onAnimationFinished(id: String?)
    }
}

File: platforms/ios/Sources/Trace/TraceView.swift
// TraceView.swift — UIView that hosts a Skia surface and renders a Trace source.
// API mirrors SPEC §5.2.

import UIKit
import SkiaKit

public protocol TraceViewDelegate: AnyObject {
    func traceView(_ view: TraceView, didTap elementId: String, at point: CGPoint)
    func traceView(_ view: TraceView, didFinishAnimation forElementId: String?)
}

public final class TraceView: UIView {
    public enum Source {
        case file(URL)
        case data(Data)
        case session(TraceSession)
    }

    public var source: Source? {
        didSet { reload() }
    }
    public weak var delegate: TraceViewDelegate?

    public var highlightedIds: Set<String> = [] {
        didSet { requestRedraw() }
    }
    public var highlightColor: UIColor = .systemOrange { didSet { requestRedraw() } }
    public var highlightStrokeBoost: CGFloat = 1.5

    private let renderer = TraceRenderer()
    private var animator: Animator?
    private var skiaView: SkiaView!
    private var traceFile: TraceFile?
    private var traceSession: TraceSession?

    public override init(frame: CGRect) {
        super.init(frame: frame); setUp()
    }
    public required init?(coder: NSCoder) {
        super.init(coder: coder); setUp()
    }

    private func setUp() {
        backgroundColor = .clear
        renderer.owner = self
        skiaView = SkiaView(frame: bounds)
        skiaView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        skiaView.drawCallback = { [weak self] canvas in
            self?.drawOn(canvas)
        }
        addSubview(skiaView)
    }

    private func reload() {
        guard let source = source else {
            traceFile = nil; traceSession = nil
            skiaView.setNeedsDisplay(); return
        }
        do {
            switch source {
            case .file(let url):
                let file = try TraceFile(url: url)
                try renderer.load(from: file)
                self.traceFile = file
                self.traceSession = nil
            case .data(let data):
                let file = try TraceFile(data: data)
                try renderer.load(from: file)
                self.traceFile = file
                self.traceSession = nil
            case .session(let session):
                self.traceSession = session
                self.traceFile = try TraceFile(data: try session.snapshot())
                try renderer.load(from: self.traceFile!)
            }
            startOrResetAnimator()
            skiaView.setNeedsDisplay()
        } catch {
            NSLog("[Trace] load failed: \(error)")
        }
    }

    private func drawOn(_ canvas: SkCanvas) {
        guard let file = traceFile else { return }
        // Apply viewBox → view transform once per frame.
        let vb = file.viewBox
        let sx = bounds.width / vb.width
        let sy = bounds.height / vb.height
        let s = min(sx, sy)
        let dx = (bounds.width - vb.width * s) * 0.5 - vb.minX * s
        let dy = (bounds.height - vb.height * s) * 0.5 - vb.minY * s
        canvas.save()
        canvas.translate(dx, dy)
        canvas.scale(s, s)
        renderer.draw(on: canvas, highlighted: highlightedIds,
                      highlightColor: highlightColor, boost: highlightStrokeBoost)
        canvas.restore()
    }

    func requestRedraw() { skiaView.setNeedsDisplay() }

    private func startOrResetAnimator() {
        animator?.stop()
        animator = Animator(renderer: renderer, file: traceFile!, session: traceSession) { [weak self] id in
            guard let self = self else { return }
            self.delegate?.traceView(self, didFinishAnimation: id)
        }
        animator?.start()
    }

    // MARK: - Hit-testing

    public override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        guard let touch = touches.first, let file = traceFile else { return }
        let viewPt = touch.location(in: self)
        let vb = file.viewBox
        let sx = bounds.width / vb.width
        let sy = bounds.height / vb.height
        let s = min(sx, sy)
        let dx = (bounds.width - vb.width * s) * 0.5 - vb.minX * s
        let dy = (bounds.height - vb.height * s) * 0.5 - vb.minY * s
        // Inverse of the draw transform — map screen to viewBox.
        let vbX = (viewPt.x - dx) / s
        let vbY = (viewPt.y - dy) / s
        // Hit-test runs in Rust. Skia is never queried for identity.
        if let id = file.hitTest(x: vbX, y: vbY, mode: .visiblePainted) {
            delegate?.traceView(self, didTap: id, at: viewPt)
        }
    }
}
